src/Random.mo

```diff
@@ -8,63 +18,206 @@ module {
 
   public let blob : shared () -> async Blob = rawRand;
 
-  /// Opinionated choice of PRNG from a given seed.
-  public func new(seed : Blob) : Random {
-    todo()
+  /// Creates a fast pseudo-random number generator using the SFC64 algorithm.
+  /// This provides statistical randomness suitable for simulations and testing,
+  /// but should not be used for cryptographic purposes.
+  /// The seed blob's first 8 bytes are used to initialize the PRNG.
+  public func fast(seed : Nat64) : Random {
+    let prng = PRNG.sfc64a();
+    prng.init(seed);
+    Random(
+      func() {
+        // Generate 8 bytes directly from a single 64-bit number
+        let n = prng.next();
+        let bytes = VarArray.repeat<Nat8>(0, 8);
+        bytes[0] := Nat8.fromNat(Nat64.toNat(n & 0xFF));
+        bytes[1] := Nat8.fromNat(Nat64.toNat((n >> 8) & 0xFF));
+        bytes[2] := Nat8.fromNat(Nat64.toNat((n >> 16) & 0xFF));
+        bytes[3] := Nat8.fromNat(Nat64.toNat((n >> 24) & 0xFF));
+        bytes[4] := Nat8.fromNat(Nat64.toNat((n >> 32) & 0xFF));
+        bytes[5] := Nat8.fromNat(Nat64.toNat((n >> 40) & 0xFF));
+        bytes[6] := Nat8.fromNat(Nat64.toNat((n >> 48) & 0xFF));
+        bytes[7] := Nat8.fromNat(Nat64.toNat((n >> 56) & 0xFF));
+        Blob.fromArray(Array.fromVarArray(bytes))
+      }
+    )
   };
 
-  /// Uses entropy from the management canister with automatic resupply.
-  public func newAsync() : AsyncRandom {
-    todo()
+  /// Creates a random number generator suitable for cryptography
+  /// using entropy from the ICP management canister with automatic resupply.
+  public func crypto() : AsyncRandom {
+    AsyncRandom(func() : async* Blob { await rawRand() })
   };
 
   public class Random(generator : () -> Blob) {
+    var iter : Iter.Iter<Nat8> = Iter.empty();
+    let bitIter : Iter.Iter<Bool> = object {
+      var mask = 0x00 : Nat8;
+      var byte = 0x00 : Nat8;
+      public func next() : ?Bool {
+        if (0 : Nat8 == mask) {
+          switch (iter.next()) {
+            case null { null };
+            case (?w) {
+              byte := w;
+              mask := 0x40;
+              ?(0 : Nat8 != byte & (0x80 : Nat8))
+            }
+          }
+        } else {
+          let m = mask;
+          mask >>= (1 : Nat8);
+          ?(0 : Nat8 != byte & m)
+        }
+      }
+    };
 
     /// Random choice between `true` and `false`.
     public func bool() : Bool {
-      todo()
+      switch (bitIter.next()) {
+        case (?bit) { bit };
+        case null {
+          iter := generator().vals();
+          switch (bitIter.next()) {
```

Comment:
```suggestion
          switch (bitIter.next()) {
```
Lucky we never specified that an iterator cannot produce ?v after already having produced null. I guess the alternative would be to assign a new bitIter, but I think this fine...

Reply:
That's a good point. We would definitely want to change this if it was exposed in the public API. This was directly taken from the original Random implementation ([here](https://github.com/dfinity/motoko-base/blob/9709055b7d81b51e35ddfbba0fae9efe5d52188a/src/Random.mo#L78)).